reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt

```diff
@@ -0,0 +1,270 @@
+import com.here.ort.utils.toHexString
+
+import java.io.File
+import java.nio.file.Path
+import java.security.MessageDigest
+
+/**
+ * Creates a notice file containing the licenses for all non-excluded projects and packages, listed by their identifier.
+ * Detected copyrights are listed next to the licenses.
+ *
+ * If a [FileArchiver] is configured for the scanner the reporter will try to download any archived license files and
+ * include them in the notice file. Licenses detected in those license files will not be listed again, instead all
+ * copyrights associated to those licenses will be listed below the license file's content.
+ */
+class NoticeByPackageReporter : AbstractNoticeReporter() {
+    override val reporterName = "NoticeByPackage"
+    override val defaultFilename = "NOTICE_BY_PACKAGE"
+
+    override fun createProcessor(input: ReporterInput): NoticeProcessor = NoticeByPackageProcessor(input)
+}
+
+class NoticeByPackageProcessor(input: ReporterInput) : AbstractNoticeReporter.NoticeProcessor(input) {
+    companion object {
+        val DEFAULT_ARCHIVE_DIR by lazy { getUserOrtDirectory().resolve("scanner/archive") }
+        const val LICENSE_SEPARATOR = "\n  --\n\n"
+    }
+
+    override fun process(data: AbstractNoticeReporter.NoticeReportData): List<() -> String> =
+        mutableListOf<() -> String>().apply {
+            add { data.headers.joinToString(AbstractNoticeReporter.NOTICE_SEPARATOR) }
+
+            if (data.headers.isNotEmpty()) {
+                add { AbstractNoticeReporter.NOTICE_SEPARATOR }
+            }
+
+            val projectIds = input.ortResult.getProjects().map { it.id }
+            val projectFindings = data.findings
+                .filter { (id, _) -> id in projectIds }
+                .filter { (_, licenseFindingsMap) -> licenseFindingsMap.isNotEmpty() }
+            val packageFindings = data.findings
+                .filterNot { (id, _) -> id in projectIds }
+
+            addProjectFindings(projectFindings)
+
+            if (packageFindings.isNotEmpty() && projectFindings.isNotEmpty()) {
+                add { AbstractNoticeReporter.NOTICE_SEPARATOR }
+            }
+
+            if (packageFindings.isEmpty()) {
+                add { data.headerWithoutLicenses }
+            } else {
+                add { data.headerWithLicenses }
+
+                addPackageFindings(packageFindings)
+            }
+
+            data.footers.forEach { footer ->
+                add { AbstractNoticeReporter.NOTICE_SEPARATOR }
+                add { footer }
+            }
+        }
+
+    private fun MutableList<() -> String>.addProjectFindings(findings: Map<Identifier, LicenseFindingsMap>) {
+        if (findings.isNotEmpty()) {
+            add { "This software includes external packages and source code.\n" }
+            add { "The applicable license information is listed below:\n" }
+        }
+
+        findings.forEach { (_, licenseFindingsMap) ->
```

Comment:
What's the reason for adding separate entries for each project rather than squashing them.
A NOTICE file targets always exactly one release (artifact) and thus the project structures can be considered as implementation detail. E.g. how fine granular does one want to split-up the project.

Some projects even use obfuscation which effectively hides that internal structure - while it then gets exposed through the notices. What do you think about squashing the entries instead?

Reply:
Good point, I think this would not have exposed any structure, because no identifiers are printed to the output, but the same licenses could have been listed multiple times. I am merging the maps now, and added a separate commit to not repeat the merge logic.

Reply:
Thanks!